### DISCUSSION TITLE
the traffic light doesn't work

### DISCUSSION THREAD

**Discussion by rufusms** [micbay/RaceTimerAndController#25] [Q&A] [answered]
Hello, I want to comment on the problem I have with the traffic light, I have tried everything, but I have not been able to get it to work, the circuit is basically a quarter of a 4-in-1 display module, which they sell on Aliexpress (https://es.aliexpress. com/w/wholesale-led-matrix-8x8-max7219.html), to which I have plugged the perfboard with the LEDs according to the scheme adapted to the card, I have checked it many times and the connections are fine, if I put it alone With a test program it works well but connected with others it does not, I have even tried replacing the max 7219 with an original one, but the result is the same, I attach a video so you can see what it does, any ideas? They don't fit me.

https://github.com/user-attachments/assets/ef98dc0c-cd12-4b73-992d-9c847aa07c6b

**Comment by micbay** [comment]
Along with double checking the wiring, also, double check that the start lights are addressed correctly in the code.
By default, the controller will assume the startlight address is 1 greater than the number of lanes set by the 'LANE_COUNT' settings variable. For 2 lanes, makes sure your LANE_COUNT value assigned in the defaultSettings.h or in localSettings.h, is set to 2. The `LED_BAR_COUNT = LANE_COUNT +1` and is used to initialize the LED chain object `lc` at line 192
```
LedControl lc = LedControl(PIN_TO_LED_DIN, PIN_TO_LED_CLK, PIN_TO_LED_CS, LED_BAR_COUNT);
```
And then to initialize the LED bar and startlight state in `setup()` at line 1872
```
  // --- SETUP LED 7-SEG, 8-DIGIT MAX7219 LED Globals ------
  // Initialize all the displays
  for(int deviceID = 0; deviceID < LED_BAR_COUNT; deviceID++) {
    // The MAX72XX is in power-saving mode on startup
    lc.shutdown(deviceID, false);
    // intensity range from 0-15, higher = brighter
    lc.setIntensity(deviceID, 8);
    // Blank the LED digits
    lc.clearDisplay(deviceID);
  }
```
Another thing you might try is to directly power the start light, rather than run power through the full set of daisy chained LED displays. The signals must still be daisy chained, but the V+ and GND can work better sometimes going directly to each device.

All that said, with the above issues I would still expect to see the start lights changing individually, which yours do not, so my first impression is that addressing and noise on the power line are not the issue.

All the lights appear to turn on and off together as one LED are you positive the wiring has not accidentally swapped the LED +/- with the MAX7219 seg vs dig inputs? When you say that they work with a test program, does that test program individually turn on and off the lights in proper sequence? Does your test program involve the LED timer bars in the chain?

**Reply by rufusms** [reply, marked as answer]
At last !!!! , it already works. Yesterday you suggested that I review the connections etc..., after reviewing everything and not finding any fault, I decided to change the 7 segment display modules, I put in two new ones, I always ask for more, and I connected everything again with the program and then I saw that it started to work, this has been a little confusing, because I have done many, many tests from the beginning with the same results, I have attached a video, thank you very much again for this system and the many explanations of how it works, thank you.

https://github.com/user-attachments/assets/95a0525f-a392-42d9-a5de-1ff16a8541c9

**Comment by micbay** [comment]
Awesome! It certainly gets tricky if one of the device or chips are defective. Perhaps the no-op register was bad on one of the LED 7-segment modules, or the data out lines were not working properly. Sometimes the device gets stuck in an odd state when programming as well. Disconnecting all power and programming lines then powering up via external power can clear odd things that can sometimes happen after uploading a sketch update.
Regardless, glad you got it worked out, the build looks great!
